pytest/tests/sanity/staking_single_shard.py

```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/env python3
+# Spins up 5 validating nodes. Let validators track a single shard.
+# Add a dumper node for the state sync headers.
+# Add an RPC node to issue tx and change the state.
+# Send random transactions between accounts in different shards.
+# Shuffle the shard assignment of validators and check if they can sync up.
+
+from collections import defaultdict
+import unittest
+import sys
+import pathlib
+
+sys.path.append(str(pathlib.Path(__file__).resolve().parents[2] / 'lib'))
+
+from transaction import (sign_staking_tx)
+
+from configured_logger import logger
+from cluster import start_cluster
+import state_sync_lib
+from utils import wait_for_blocks
+import simple_test
+
+EPOCH_LENGTH = 10
+
+NUM_VALIDATORS = 5
+
+# Shard layout with 5 roughly equal size shards for convenience.
+SHARD_LAYOUT = {
+    "V1": {
+        "boundary_accounts": [
+            "fff",
+            "lll",
+            "rrr",
+        ],
+        "version": 2,
+        "shards_split_map": [],
+        "to_parent_shard_map": [],
+    }
+}
+
+NUM_SHARDS = len(SHARD_LAYOUT["V1"]["boundary_accounts"]) + 1
+
+ALL_ACCOUNTS_PREFIXES = [
+    "aaa",
+    "ggg",
+    "lll",
+    "sss",
+]
+
+
+class StateSyncValidatorShardSwap(unittest.TestCase):
+
+    def _prepare_cluster(self,
+                         with_rpc=False,
+                         shuffle_shard_assignment=False,
+                         centralized_state_sync=False,
+                         decentralized_state_sync=False):
+        (node_config_dump,
+         node_config_sync) = state_sync_lib.get_state_sync_configs_pair(
```

Comment:
All of these tests are using the same config, one in which it is specified that an external source is available in state_part_dir. In 2.3.0-rc.4 we made it so that if an external source is configured, it will be used directly and no requests will be made to peers. Then in `test_state_sync_with_restaking_decentralized_only` you have the same config but you don't dump state parts in the state_part_dir. Hence the node just gets stuck looking for state parts in the external directory.